Toggl.Networking/Sync/Push/Request.cs

```diff
@@ -1,99 +1,183 @@
+        public List<PushAction> TimeEntries { get; set; }
 
-        public SingletonUpdateAction<IPreferences> Preferences { get; set; }
-        public SingletonUpdateAction<IUser> User { get; set; }
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public List<PushAction> Tags { get; set; }
+
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public List<PushAction> Projects { get; set; }
+
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public List<PushAction> Clients { get; set; }
+
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public List<PushAction> Tasks { get; set; }
+
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public List<PushAction> Workspaces { get; set; }
+
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public PushAction Preferences { get; set; }
+
+        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+        public PushAction User { get; set; }
 
         public Request(string userAgent)
         {
             Ensure.Argument.IsNotNullOrWhiteSpaceString(userAgent, nameof(userAgent));
             this.userAgent = userAgent;
         }
 
+        [JsonIgnore]
         public bool IsEmpty
-            => TimeEntries.None()
-            && Tags.None()
-            && Projects.None()
-            && Clients.None()
-            && Tasks.None()
-            && Workspaces.None()
+            => (TimeEntries == null || TimeEntries.None())
+            && (Tags == null || Tags.None())
+            && (Projects == null || Projects.None())
+            && (Clients == null || Clients.None())
+            && (Tasks == null || Tasks.None())
+            && (Workspaces == null || Workspaces.None())
             && Preferences == null
             && User == null;
 
         public void CreateTimeEntries(IEnumerable<ITimeEntry> timeEntries)
         {
-            timeEntries
+            var list = timeEntries.ToList();
+            if (list.Count == 0)
+                return;
+
+            if (TimeEntries == null)
+            {
+                TimeEntries = new List<PushAction>();
+            }
+
+            list
                 .Select(timeEntry => new TimeEntry(timeEntry) { CreatedWith = userAgent })
-                .Select(timeEntry => new CreateAction<TimeEntry>(timeEntry))
+                .Select(timeEntry => PushAction.Create(timeEntry.Id, timeEntry))
                 .AddTo(TimeEntries);
         }
 
         public void UpdateTimeEntries(IEnumerable<ITimeEntry> timeEntries)
         {
-            timeEntries
-                .Select(bte => new UpdateAction<ITimeEntry>(bte))
+            var list = timeEntries.ToList();
+            if (list.Count == 0)
+                return;
+
+            if (TimeEntries == null)
+            {
+                TimeEntries = new List<PushAction>();
+            }
```

Comment:
This can be used on all of similar places in this file.

```suggestion
            TimeEntries ??= new List<PushAction>();
```